README.md
# mstm-spectrum
## About
Python wrapper for multiple sphere T-matrix (MSTM) code to calculate surface plasmon resonance (SPR) spectrum and *fit* it to experiment.

Based and inspired by MSТM GUI code <https://git.stim.ee.uh.edu/optics/mstm-gui.git> by David Mayerich.

Multi-Sphere T-Matrix code is proposed by Dr. Daniel Mackowski and Dr. Michael Mishchenko in:
"A multiple sphere T-matrix Fortran code for use on parallel computer clusters,"
*Journal of Quantitative Spectroscopy and Radiative Transfer* (2011) 112 2182-2192
<https://dx.doi.org/10.1016/j.jqsrt.2011.02.019>.

Please cite the above reference if using MSTM code.

### Dependencies

* **Python** (tested: Python2.7 & 3.4 under Debian Linux; Anaconda distributed Python2 under Windows)
* **NumPy** - numerical python library
* **SciPy** - scientific python library (Powell minimizer)
* **MatPlotLib** - plotting with python (optional)
* **tkinter**, **PIL** - tk libraries and Python image libray - for GUI (optional)

### Installation

* download and unpack [latest release](releases/latest) archive (MSTM binaries for Linux Debian x64 and Windows x32 are included)

or

* clone this repository but use your own MSTM binaries (downloaded or compiled)

### Contributors

* Avakyan L.A. <laavakyan@sfedu.ru>
* Pryadchenko V.V. <vvpryadchenko@sfedu.ru>

For details please refer to source files.

## Usage

### Graphical user interface

The intuitive (hope, it is) graphical user interface
 can be run executing script `mstm_studio_support.py` or `mstm_studio.py`.

![GUI screenshot image][screen_gui]

### Python scripting

Alternatively, the python scripting way may be used.
This is less intuitive, but allows fine tuning of the calculations.

The possible workflow is:

1. Place files in the same directory:
    1. experiment file,
    1. binaries named as:
        * for Windows: `mstm.exe`
        * for Linux: `run_mstm.sh` script to run `mstm.x` (see example)

        Source code and binaries can be obtained on [MSTM website](http://eng.auburn.edu/users/dmckwski/scatcodes/)
        or using [direct download link](http://eng.auburn.edu/users/dmckwski/scatcodes/mstm%20v3.0.zip).
        (You may contact us if suffer compilation problems)
1. Edit `start_fit.py` file to suit your needs. The supplied file contains:
    1. path to the directory with the scripts (remove these lines if scripts are stored in current directory):

        ``` python
        import sys
        # set the path to mstm_spectrum scripts. Binary mstm files should be in current folder.
        sys.path.append('/home/leon/ltg_projects/fit-T-matrix/mstm-spectrum')
        ```
    1. names of imported modules

        ``` python
        from mstm_spectrum import ExplicitSpheres
        from alloy_AuAg import AlloyAuAg
        from fit_spheres_optic import Fitter, FixConstraint
        ```
    1. setup of the experiment file name, background contribution and surrounding material:

        ``` python
        fitter = Fitter('example/optic_sample19.dat')
        fitter.set_background('lorentz')  # 'constant', 'linear' or 'lorentz'
        fitter.set_matrix('glass')  # 'glass', 'water', 'air' or explicit value, i.e. 1.66+0.1j
        ```
    1. specification the initial configuration of spheres.
       In this example, we will start from 3 silver spheres put in the XY plane:

        ``` python
        #                              x             y            z
        spheres = ExplicitSpheres(3, [-10, 10, 0], [0, 0, 14], [0, 0, 0],
                                  [12, 12, 12], [AlloyAuAg(x_Au=0.0)]*3)
        fitter.set_spheres(spheres)
        ```
    1. don't vary position of one of the spheres to save some computational resourses:

        ``` python
        fitter.add_constraint(FixConstraint('x0'))
        fitter.add_constraint(FixConstraint('y0'))
        fitter.add_constraint(FixConstraint('z0'))
        ```
    1. show some info, run, and report results:

        ``` python
        fitter.report_freedom()
        raw_input('Press enter')

        fitter.run()

        fitter.report_result()
        raw_input('Press enter')
        ```
    1. execution of this script will show some stats:

        ```
        Number of spheres:      3
        Background parameters:  3
        Degree of freedom
        internal:       4
        external:       9

        ```
        After pressing enter (required by `raw_input` command),
        the fitting will start.
        At each accepted minimizing step the plot will be updated:

        ![Screenshot image][screen]


## Citation

If you used this code in a scientific paper please cite original MSTM code reference and the following reference:

L. Avakyan, M. Heinz, A. Skidanenko, K. Yablunovskiy, J. Ihlemann, J. Meinertz, C. Patzig, M. Dubiel, L. Bugaev
*J. Phys.: Condens. Matter* (2018) 30 045901 [[doi](http://doi.org/10.1088/1361-648X/aa9fcc)]


[screen_gui]: images/screenshot-gui.jpg?raw=true "GUI screenshot"
[screen]: images/screenshot-example.jpg?raw=true "Screenshot of example run"


CHANGES.txt


2017-12-16

* Implemented GUI using Tk backend. Activated when `mstm_studio.py` or `mstm_studio_support.py` script is run.

* Implemented analytical AuAg alloy dielectric function (refraction indeces).


2017-11-04 

* Fit procedure is split in two steps now. In internal (fast) step the scale and background parameters are varied, so that no recalculation of the spectra required.

2019-04-15

* Fixed undefined behaviour originated from ordering of keys in dictionary containing varied parameters 

* Extended capabilities for fitting of non-MSTM contributions


fit_spheres_optic.py
# -*- coding: utf-8 -*-
#-----------------------------------------------------#
#                                                     #
# This code is a part of T-matrix fitting project     #
# Contributors:                                       #
#  L. Avakyan <laavakyan@sfedu.ru>                    #
#  A. Skidanenko <ann.skidanenko@ya.ru>               #
#                                                     #
#-----------------------------------------------------#
"""
  Fitting of particle aggreagate T-matrix spectrum
  to experimental SPR spectrum.
"""
from __future__ import print_function
import os
from mstm_spectrum import SPR, ExplicitSpheres
from contributions import ConstantBackground
import numpy as np
from scipy import interpolate
import scipy.optimize as so
from datetime import datetime
import matplotlib.pyplot as plt
import threading

# use input in both python2 and python3
try:
   input = raw_input
except NameError:
   pass
# use xrange in both python2 and python3
try:
    xrange
except NameError:
    xrange = range

class Parameter(object):
    """
    Class for parameter object used for storage of
    parameter's name, value and variation limits.

    Parameter naming conventions:
    scale - multiplier for MSTM spectra
    ext%i - extra parameter, like background, peaks or mie contributions
    a%i - sphere radius
    x%i, y%i, z%i - coordinates of sphere center
    ...
    where %i - number (0, 1, ...)
    """
    def __init__(self, name, value=1, min=None, max=None, internal_loop=False):
        """
        Parameter object

        name : str
            name of parameter used for constraints etc
        value : float
            initial value of parameter
        min, max : float
            bounds for parameter variation (optional)
        internal_loop : bool
            if True the parameter will be allowed to vary in internal (fast)
            loop, which does not require recalculation of spectrum.
            Note: this flag will be removed in future
        varied : bool
            whether should be changed or not
        """
        self.name = name
        self.value = self.ini_value = value
        self.min = min
        self.max = max
        self.internal_loop = internal_loop
        self.varied = True
        # ...

    def __str__(self):
        return '%s : %f' % (self.name, self.value)


class Constraint(object):
    def apply(self, params):
        """
        modify the params dict
        according to given constranint.
        !Abstract method!
        """
        pass


class FixConstraint(Constraint):
    def __init__(self, prm, value=None):
        """
        fix value of parameter.

        prm : str
          parameter name
        value : float
          if None than initial value will be used.
        """
        self.prm = prm.lower()
        self.value = value

    def apply(self, params):
        assert self.prm in params
        if self.value is not None:
            params[self.prm].value = self.value
        params[self.prm].varied = False


class EqualityConstraint(Constraint):
    def __init__(self, prm1, prm2):
        """
        fix two parameters being equal
        """
        self.prm1 = prm1.lower()
        self.prm2 = prm2.lower()

    def apply(self, params):
        assert self.prm1 in params
        assert self.prm2 in params
        params[self.prm2].value = params[self.prm1].value
        params[self.prm2].varied = False


class ConcentricConstraint(Constraint):
    def __init__(self, i1, i2):
        """
        two spheres with common centers
        i1 and i2 - indexes of spheres
        """
        self.constraints = [EqualityConstraint('x%i'%i1, 'x%i'%i2),
                            EqualityConstraint('y%i'%i1, 'y%i'%i2),
                            EqualityConstraint('z%i'%i1, 'z%i'%i2)]

    def apply(self, params):
        for c in self.constraints:
            c.apply(params)


class Fitter(threading.Thread):
    """
    Class to perform fit of experimental Exctinction spectrum
    """
    def __init__(self, exp_filename, wl_min=300, wl_max=800, wl_npoints=51,
                 extra_contributions=None, plot_progress=True):
        """
        Creates the Fitter object

        Parameters:
        exp_filename : str
            name of file with experimental data
        wl_min, wl_max : float (nm)
            wavelength bounds for fitting.
        wl_npoints : int
            number of wavelengths where spectra will be calcualted and compared.
        extra_contributions : list of Contribution objects
            If None, then ConstantBackground will be used.
            Assuming that first element is a background.
            If you don't want any extra contribution, set to empty list [].
        plot_progress : bool
            Show fitting progress using matplotlib.
            Should be turned off when run on parallel cluster without gui.
        """
        super(Fitter, self).__init__()
        self._stop_event = threading.Event()  # to be able to stop outside

        self.exp_filename = exp_filename
        data = np.loadtxt(self.exp_filename)    # load data
        data = data[np.argsort(data[:,0]),:]    # sort by 0th column
        if np.max(data[:,0]) < 10:      # if values are really low
            print('WARNING: Data X column is probably in mum, automatilcally rescaling to nm.')
            data[:,0] = data[:,0] * 1000

        self.wl_min = max(wl_min, data[ 0, 0])
        self.wl_max = min(wl_max, data[-1, 0])
        self.wl_npoints = wl_npoints
        print('Wavelength limits are setted to: %f < wl < %f'% (self.wl_min, self.wl_max))
        self.wls, self.exp = self._rebin(self.wl_min, self.wl_max, self.wl_npoints,
                                         data[:,0], data[:,1])
        self.params = {}             # dictionaty of parameters objects
        self.spheres = None          # object of Spheres
        self.constraints = []        # list of Constraint objects
        self.calc = np.zeros_like(self.wls)  # calculated spectrum
        self.chisq = -1              # chi square (squared residual)

        self.plot_progress = plot_progress
        if self.plot_progress:
            plt.ion()
            self.fig = plt.figure()
            ax = self.fig.add_subplot(111)
            ax.plot(self.wls, self.exp, 'ro')
            self.line1, = ax.plot(self.wls, self.calc, 'b-')
            self.fig.canvas.draw()
            #~ self.lock = threading.Lock()  # used to sync with main thread where plot
        # set scale as default
        self.set_scale()
        # add extra contributions
        self.extra_contributions = []
        self.set_extra_contributions(extra_contributions)
        # set matrix material as default
        self.set_matrix()
        # callback function supplied outside
        self._cbuser = None

    def _rebin(self, xmin, xmax, N, x, y):
        """
        hidden method used to rebin data to uniform scale
        """
        f = interpolate.interp1d(x, y)
        xnew = np.linspace(xmin, xmax, N)
        ynew = f(xnew)
        return xnew, ynew

    def _print_params(self):
        for key in sorted(self.params):
            print('params[ %s ] \t %s ' % (key, self.params[key]))

    def set_matrix(self, material='AIR'):
        """
        set refraction index of matrix material

        material : {'AIR'|'WATER'|'GLASS'} or float
            the name of material or
            refraction index value.
        """
        self.MATRIX_MATERIAL = material

    def set_scale(self, value=1):
        if 'scale' not in self.params:
            self.params['scale'] = Parameter('scale', value=value, internal_loop=True)
        else:
            self.params['scale'].value = value
            self.params['scale'].ini_value = value

    def set_extra_contributions(self, contributions, initial_values=None):
        """
        Add extra contributions and initialize corresponding params.

        initial_values : float array
        """
        # remove old parameters
        i_tot = 0
        for contribution in self.extra_contributions:
            if contribution is not None:
                n = contribution.number_of_params
                for _ in range(n):
                    self.params.pop('ext%i' % i_tot)
                    i_tot += 1

        if contributions is None:
            contributions = [ConstantBackground(self.wls, 'ConstBkg')]
        self.extra_contributions = contributions[:]

        # create new parameter objects
        n_tot = 0
        i_tot = 0
        for contribution in self.extra_contributions:
            n = contribution.number_of_params
            for _ in range(n):
                self.params['ext%i' % i_tot] = Parameter('ext%i' % i_tot, value=0.1, internal_loop=True)
                i_tot += 1
            n_tot += n
        self.extra_contrib_params_count = n_tot
        if initial_values is not None:
            assert len(initial_values) == n_tot
            for i in range(n_tot):
                if initial_values[i] is not None:
                    self.params['ext%i' % i].value = initial_values[i]
                    self.params['ext%i' % i].ini_value = initial_values[i]
        # print(self.params)

    def set_spheres(self, spheres):
        """
        specify the spheres aggregate

        spheres : mstm_spectrum.Spheres
        """
        if self.spheres is not None:  # remove parameters of old spheres
            for i in xrange(self.spheres.N):
                self.params.pop('a%i' % i)
                self.params.pop('x%i' % i)
                self.params.pop('y%i' % i)
                self.params.pop('z%i' % i)
        if spheres is not None:
            self.spheres = spheres
            for i in xrange(self.spheres.N):
                self.params['a%i' % i] = Parameter('a%i' % i, self.spheres.a[i])
                self.params['x%i' % i] = Parameter('x%i' % i, self.spheres.x[i])
                self.params['y%i' % i] = Parameter('y%i' % i, self.spheres.y[i])
                self.params['z%i' % i] = Parameter('z%i' % i, self.spheres.z[i])
        else:
            self.set_spheres(ExplicitSpheres())  # empty spheres object

    def update_spheres(self):
        """
        Set spheres radii and positions to values from params dict
        """
        assert self.spheres is not None
        for i in xrange(len(self.spheres)):
            self.spheres.a[i] = self.params['a%i' % i].value
            self.spheres.x[i] = self.params['x%i' % i].value
            self.spheres.y[i] = self.params['y%i' % i].value
            self.spheres.z[i] = self.params['z%i' % i].value

    def update_params(self, values, internal=False):
        """
        Put values from optimized to params

        internal : bool
            if True than internal variables will be updated (scale, bkg, ..)
        """
        try:  # if not iterable (single value in values)
            len(values)
        except:
            print('WARNING: values is not a list')
            values = [values]
        if internal:  # internal (fast) loop parameters
            self.params['scale'].value = values[0]
            for i in range(self.extra_contrib_params_count):
                self.params['ext%i' % i].value = values[i+1]  # 0th is scale
            print('inner: ', (self.extra_contrib_params_count+1), values)
        else:
            # apply constraints, -- up to now works only for MSTM
            for c in self.constraints:  # apply before
                c.apply(self.params)
            # update params
            i_tot = 0
            for i in range(len(self.spheres)):
                for key in ('a%i'%i,'x%i'%i,'y%i'%i,'z%i'%i):
                    if self.params[key].varied:
                        self.params[key].value = values[i_tot]
                        i_tot += 1
            assert i_tot == len(values)
            for c in self.constraints:  # and apply after
                c.apply(self.params)
            self.report_result(msg='[%s] Scale: %.3f Bkg: %.2f\n' % (str(datetime.now()),
                self.params['scale'].value, self.params['ext0'].value))  # may be verbous!

    def add_constraint(self, cs):
        """
        adds contraints on the parameters.
        Usefull for the case of core-shell and layered structures.

        cs : Contraint object or list of Contraint objects
        """
        try:
            _ = iter(cs)
        except TypeError:
            cs = [cs]
        for c in cs:
            self.constraints.append(c)

    def get_spectrum(self):
        """
        Calculate the spectrum of agglomerates using mstm_spectrum module.
        """
        if self.stopped():
            raise Exception('Fitting interrupted')

        #~ self._apply_constraints

        spr = SPR(self.wls)
        spr.environment_material = self.MATRIX_MATERIAL

        self.update_spheres()
        spr.set_spheres(self.spheres)

        result = np.zeros_like(self.wls)
        if len(self.spheres) > 0:
            #~ self.lock.acquire()
            try:
                _, extinction = spr.simulate('exct.dat')
                result = np.array(extinction)
            except Exception as e:
                print(e)
            #~ finally:
                #~ self.lock.release()

        # perform fast fit over internal variables (scale, bkg, ..)
        values_internal = []
        values_internal.append(self.params['scale'].value)
        for i in range(self.extra_contrib_params_count):
            values_internal.append(self.params['ext%i' % i].value)

        def target_func_int(values):
            """ target function for internal fit (fast loop) """
            self.update_params(values, internal=True)

            y_dat = self.exp
            assert self.params['scale'].value == values[0]
            y_fit = values[0] * result
            n_tot = 1  # scale is values[0]
            for contribution in self.extra_contributions:
                n = contribution.number_of_params
                y_fit += contribution.calculate(values[n_tot:n_tot+n])
                n_tot += n
            self.chisq = np.sum((y_fit - y_dat)**2)
            #~ self.chisq = np.sum((y_fit - y_dat)**2 * (y_dat/np.max(y_dat)+0.001)) / np.sum((y_dat/np.max(y_dat)+0.001))
            #~ self.chisq = np.sum( (y_fit - y_dat)**2 * y_dat**3 ) * 1E3
            print(self.chisq)
            return self.chisq
        #~ print('/ Internal fit loop /')
        result_int = so.minimize(fun=target_func_int, x0=values_internal, method='Nelder-Mead', tol=1E-5,
                                 options={'maxiter':100, 'disp':False})
        values_internal = result_int.x

        self.update_params(values_internal, internal=True)

        self.calc = self.params['scale'].value * result
        n_tot = 1  # 0th is scale
        for contribution in self.extra_contributions:
            n = contribution.number_of_params
            self.calc += contribution.calculate(values_internal[n_tot:n_tot+n])
            n_tot += n
        return self.calc

    def target_func(self, values):
        """ main target function """
        self.update_params(values)

        y_dat = self.exp
        y_fit = self.get_spectrum()
        self.chisq = np.sum((y_fit - y_dat)**2)
        #~ self.chisq = np.sum((y_fit - y_dat)**2 * (y_dat/np.max(y_dat)+0.001)) / np.sum((y_dat/np.max(y_dat)+0.001))
        #~ self.chisq = np.sum( (y_fit - y_dat)**2 * (y_dat + np.max(y_dat*0.001))**3)
        #~ self.chisq = np.sum( (y_fit - y_dat)**2 * y_dat**3 ) * 1E3
        #print(chisq)
        return self.chisq

    def set_callback(self, func):
        """
        set callback function which will be performed on
        each step of outer optimization loop

        func : function(values),
            values -- list of values passed from optimization routine
        """
        self._cbuser = func

    def _cbplot(self, values):
        """
        callback function
        """
        #~ self.lock.acquire()  # will wait here
        #~ try:
        #print('Scale: %0.3f Bkg: %0.3f ChiSq: %.8f'% (self.params['scale'].value,
        #      self.params['bkg0'].value, self.chisq) )
        if self._cbuser is not None:  # call user-supplied function
            self._cbuser(self, values)
        if self.plot_progress:
            self.line1.set_ydata(self.calc)
            self.fig.canvas.draw()
            #~ plt.pause(0.05)  # this lead of grabbing of the focus by the plot window
            self.fig.canvas.start_event_loop(0.05)
            #from:
            #https://stackoverflow.com/questions/45729092/make-interactive-matplotlib-window-not-pop-to-front-on-each-update-windows-7
            #~ backend = plt.rcParams['backend']
            #~ import matplotlib
            #~ if backend in matplotlib.rcsetup.interactive_bk:
                #~ figManager = matplotlib._pylab_helpers.Gcf.get_active()
                #~ if figManager is not None:
                    #~ canvas = figManager.canvas
                    #~ if canvas.figure.stale:
                        #~ canvas.draw()
                    #~ canvas.start_event_loop(0.05)
        #~ finally:
            #~ self.lock.release()
        #~ input('pe')

    def _apply_constraints(self):
        for c in self.constraints:
            c.apply(self.params)

    def run(self, tol=1E-6, maxsteps=400):
        """
        Search for the best spheres aggregate

        tol : float
            tolerance in change of residual (target) function
        maxsteps : int
            maximum number of steps of the search.
        """
        self._apply_constraints()
        # pack parameters to values
        values = []
        for i in range(len(self.spheres)):
            for key in ('a%i'%i,'x%i'%i,'y%i'%i,'z%i'%i):
                if self.params[key].varied:
                    values.append(self.params[key].value)
        # run optimizer
        result = so.minimize(fun=self.target_func, x0=values, method='Powell', tol=tol,
                             options={'maxiter':maxsteps, 'disp':True}, callback=self._cbplot)
        self.update_params(result.x)

    def stop(self):
        # https://stackoverflow.com/questions/323972/is-there-any-way-to-kill-a-thread-in-python
        self._stop_event.set()

    def stopped(self):
        return self._stop_event.is_set()

    def report_freedom(self):
        self._apply_constraints()
        N = len(self.spheres)
        s = 'Number of spheres:\t%i\n' % N
        n_tot = 0
        for contribution in self.extra_contributions:
            n = contribution.number_of_params
            s += 'Extra contrib. %s with %i params\n' % (contribution.name, n)
            n_tot += n
        s += 'Total number of extra params:\t%i\n' % n_tot
        n_fast = n_slow = n_fix = 0
        for key in self.params:
            if self.params[key].varied:
                if self.params[key].internal_loop:
                    n_fast += 1
                else:
                    n_slow += 1
            else: # not varied
                n_fix += 1
        assert n_fast + n_slow + n_fix == 1 + 4*N + n_tot
        s += 'Degrees of freedom\n'
        s += '\tfast loop:\t%i\n' % n_fast
        s += '\tslow loop:\t%i\n' % n_slow
        print(s)
        return s

    def report_result(self, msg=None):
        """
        report the final values of parameters to stdout
        """
        s = 'ChiSq:\t%f\n' % self.chisq
        if msg is None:
            s += 'Optimal parameters'
        else:
            s += msg
        for key in sorted(self.params):
            s += '\n\t%s:\t%f\t(Varied:%s)\n' % (key, self.params[key].value, str(self.params[key].varied))
        print(s)
        return s

if __name__ == '__main__':
    fitter = Fitter('example/optic_sample22.dat')
    # test Mie fit
    from contributions import LinearBackground, MieSingleSphere, MieLognormSpheres
    from mstm_spectrum import Material
    fitter.set_extra_contributions([LinearBackground(fitter.wls, 'lin bkg'),
                                    MieLognormSpheres(fitter.wls, 'LN Mie')],
                                    [0.02, -0.001,
                                    0.1, 1.5, 0.5])
                                    #~ MieSingleSphere(fitter.wls, 'Mie')],
                                    #~ [0.02, -0.001,
                                    #~ 0.1, 10])
    fitter.extra_contributions[1].set_material(Material('etaGold.txt'), 1.66)
    fitter.extra_contributions[1].plot([0.1, 1.5, 0.5])
    fitter.extra_contributions[1].plot_distrib([0.1, 1.5, 0.5])
    #~ fitter.extra_contributions[1].plot([0.1, 10])
    fitter.set_spheres(None)  # no spheres, no mstm runs
    fitter.report_freedom()
    input('Press enter to run peak fitting')
    fitter.run()
    fitter.report_result()
    input('Press enter to continue')
    # test peak fit
    from contributions import LinearBackground, LorentzPeak
    fitter.set_extra_contributions([LinearBackground(fitter.wls, 'lin bkg'),
                                    LorentzPeak(fitter.wls, 'lorentz peak')],
                                    [0.02, -0.001,
                                    100, 550, 50])
    # fitter.extra_contributions[1].plot([100, 550, 50])
    fitter.set_spheres(None)  # no spheres, no mstm runs
    fitter.report_freedom()
    input('Press enter to run peak fitting')
    fitter.run()
    fitter.report_result()
    input('Press enter to continue')
    # test MSTM fit
    fitter.set_matrix('glass')
    fitter.set_extra_contributions([LinearBackground(fitter.wls, 'lin bkg')], [0.02, -0.001])
    #                         N    X      Y      Z    radius    materials
    spheres = ExplicitSpheres(2, [-1,1], [-2,2], [-3,3], [14,20], ['etaGold.txt', 'etaSilver.txt'])
    fitter.set_spheres(spheres)
    fitter.add_constraint(EqualityConstraint('x0', 'x1'))
    fitter.add_constraint(EqualityConstraint('y0', 'y1'))
    fitter.add_constraint(EqualityConstraint('z0', 'z1'))
    #~ fitter.add_constraint(FixConstraint('x0'))
    #~ fitter.add_constraint(FixConstraint('y0'))
    #~ fitter.add_constraint(FixConstraint('z0'))
    fitter.add_constraint(FixConstraint('x0', 0))
    fitter.add_constraint(FixConstraint('y0', 0))
    fitter.add_constraint(FixConstraint('z0', 0))
    fitter.report_freedom()
    input('Press enter to run MSTM fitting')

    fitter.run()
    #~ fitter.start()  # thread method
    #~ fitter.join()   # wait till end
    fitter.report_result()
    #fitter.plot_result()
    #~ y_fit = get_spectrum( wavelengths, values )
    #~ plt.plot( wavelengths, exp, wavelengths, y_fit )
    #~ #plt.axis([0, 1, 1.1*np.amin(s), 2*np.amax(s)])
    #~ plt.xlabel('Wavelength, nm')
    #~ plt.ylabel('Exctinction, a.u.')
    #~ plt.show()
    input('Press enter to finish')
    print('It is over.')
